internal/repository/user.go

```diff
@@ -0,0 +1,122 @@
+package repository
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/dilyara4949/flight-booking-api/internal/domain"
+	"github.com/google/uuid"
+)
+
+type userRepository struct {
+	db             *sql.DB
+	contextTimeout time.Duration
+}
+
+func NewUserRepository(db *sql.DB, timeout time.Duration) domain.UserRepository {
+	return &userRepository{db: db, contextTimeout: timeout}
+}
+
+var (
+	ErrUserNotFound   = errors.New("user not found")
+	ErrNothingChanged = errors.New("nothing changed")
+)
+
+const (
+	createUser  = "insert into users (id, email, password, phone, role) values ($1, $2, $3, $4, 'user') returning created_at, updated_at"
+	getUser     = "select id, email, phone, created_at, updated_at from users where id = $1;"
+	updateUser  = "update users set email = $2, phone = $3, updated_at = CURRENT_TIMESTAMP where id = $1;"
+	deleteUsers = "delete from users where id = $1"
+	getAllUsers = "select id, email, phone, created_at, updated_at from users limit $1 offset $2;"
+)
+
+func (repo *userRepository) Create(ctx context.Context, user *domain.User, password string) error {
+	ctx, cancel := context.WithTimeout(ctx, repo.contextTimeout)
+	defer cancel()
+
+	user.ID = uuid.New().String()
+
+	if err := repo.db.QueryRowContext(ctx, createUser, user.ID, user.Email, password, user.Phone).Scan(&user.CreatedAt, &user.UpdatedAt); err != nil {
+		return fmt.Errorf("create user error: %v", err)
+	}
+	return nil
+}
+
+func (repo *userRepository) Get(ctx context.Context, id string) (*domain.User, error) {
+	ctx, cancel := context.WithTimeout(ctx, repo.contextTimeout)
+	defer cancel()
+
+	row := repo.db.QueryRowContext(ctx, getUser, id)
+
+	user := domain.User{}
+
+	err := row.Scan(&user.ID, &user.Email, &user.Phone, &user.CreatedAt, &user.UpdatedAt)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ErrUserNotFound
+		}
+		return nil, fmt.Errorf("get user error: %v", err)
+	}
+
+	return &user, nil
+}
+
+func (repo *userRepository) Update(ctx context.Context, user domain.User) error {
+	ctx, cancel := context.WithTimeout(ctx, repo.contextTimeout)
+	defer cancel()
+
+	res, err := repo.db.ExecContext(ctx, updateUser, user.ID, user.Email, user.Phone)
+	if err != nil {
+		return fmt.Errorf("update user error: %v", err)
+	}
+
+	if cnt, _ := res.RowsAffected(); cnt != 1 {
+		return ErrNothingChanged
+	}
+	return nil
+}
+
+func (repo *userRepository) Delete(ctx context.Context, id string) error {
+	ctx, cancel := context.WithTimeout(ctx, repo.contextTimeout)
+	defer cancel()
```

Comment:
is there any reason to use ctx with timeout?

Reply:
why not? i thought it may be necessary in the future

Reply:
you have a request context, so when it will be cancelled execution of this logic will be cancelled too. So there is no reason to use cancellation

Reply:
got you now, i am using ctx WithTimeout there, since i dont have any timeouts in upper layers